lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/AffixCondition.java

```diff
@@ -0,0 +1,214 @@
+   */
+  static String uniqueKey(AffixKind kind, String strip, String condition) {
+    if (".".equals(condition)
+        || kind == PREFIX && strip.startsWith(condition)
+        || kind == SUFFIX && strip.endsWith(condition) && !isRegexp(condition)) {
+      return ALWAYS_TRUE_KEY;
+    }
+    return kind == SUFFIX ? ".*" + condition : condition + ".*";
+  }
+
+  /**
+   * Analyzes the given affix kind, strip and condition and returns an object able to efficiently
+   * check that condition.
+   */
+  static AffixCondition compile(AffixKind kind, String strip, String condition, String line) {
+    String stemCondition = removeStrip(strip, condition, line, kind);
+    if (stemCondition == null) {
+      return ALWAYS_FALSE;
+    }
+
+    if (stemCondition.isEmpty()) {
+      return ALWAYS_TRUE;
+    }
+
+    if (!isRegexp(stemCondition)) {
+      return substringCondition(kind, stemCondition);
+    }
+
+    try {
+      return regexpCondition(kind, escapeDash(stemCondition));
+    } catch (IllegalArgumentException e) {
+      throw new IllegalArgumentException("On line: " + line, e);
+    }
+  }
+
+  private static boolean isRegexp(String condition) {
+    return condition.contains("[") || condition.contains(".") || condition.contains("-");
+  }
+
+  /** Removes the "strip" from "condition", to check only the remaining stem part */
+  private static String removeStrip(String strip, String condition, String line, AffixKind kind) {
+    if (!isRegexp(condition)) {
+      if (kind == SUFFIX && condition.endsWith(strip)) {
+        return condition.substring(0, condition.length() - strip.length());
+      }
+      if (kind == PREFIX && condition.startsWith(strip)) {
+        return condition.substring(strip.length());
+      }
+    }
+
+    List<String> charPatterns = parse(condition);
+    try {
+      if (charPatterns.size() < strip.length()) {
+        String regexp = unitePatterns(charPatterns);
+        return strip.matches(kind == SUFFIX ? ".*" + regexp : regexp + ".*") ? "" : null;
+      }
+
+      int stripRangeStart = kind == PREFIX ? 0 : charPatterns.size() - strip.length();
+      int stripRangeEnd = kind == PREFIX ? strip.length() : charPatterns.size();
+      if (!strip.isEmpty()
+          && !strip.matches(unitePatterns(charPatterns.subList(stripRangeStart, stripRangeEnd)))) {
+        return null;
+      }
+
+      if (kind == PREFIX) {
+        return unitePatterns(charPatterns.subList(stripRangeEnd, charPatterns.size()));
+      }
+      return unitePatterns(charPatterns.subList(0, stripRangeStart));
+    } catch (PatternSyntaxException e) {
+      throw new IllegalArgumentException("On line " + line, e);
+    }
+  }
+
+  /** Produces a regexp from groups returned by {@link #parse} */
+  private static String unitePatterns(List<String> charPatterns) {
+    return charPatterns.stream()
+        .map(s -> s.length() == 1 && "()?$^{}*+|\\".indexOf(s.charAt(0)) >= 0 ? "\\" + s : s)
+        .collect(Collectors.joining());
+  }
```

Comment:
what is this? is it doing a union? For a union, it would be better to use BasicOperations.union on Automaton objects rather than mess around with regexes as strings

Reply:
It's concatenation. Maybe `unite` isn't such a great name after all. Would `join` be better?

Reply:
I see. I think I just get worried when I see escaping for `RegExp` syntax. Maybe as a `TODO` for the future we can avoid using RegExp class at all here? We could perhaps skip the parsing step and build the `Automaton` we need programmatically with `org.apache.util.automaton` APIs?